pkg/trace/stats/sublayers.go

```diff
@@ -34,6 +43,181 @@ func (v SublayerValue) GoString() string {
 	return v.String()
 }
 
+// SublayerCalculator holds arrays used to compute sublayer metrics.
+// Re-using arrays reduces the number of allocations
+type SublayerCalculator struct {
+	activeSpans      []int
+	activeSpansIndex []int
+	openSpans        []bool
+	nChildren        []int
+	execDuration     []float64
+	parentIdx        []int
+	timestamps       timestampArray
+	maxSpans         int
+}
+
+// NewSublayerCalculator returns a new SublayerCalculator used to compute sublayer metrics
+func NewSublayerCalculator() *SublayerCalculator {
+	c := &SublayerCalculator{}
+	c.initFields(defaultSublayersCalculatorMaxSpans)
+	return c
+}
+
+// initFields initialized all arrays of the sublayer calculator
+// it should be called every time we receive a trace with more than maxSpans spans
+func (c *SublayerCalculator) initFields(maxSpans int) {
+	c.maxSpans = maxSpans
+	c.activeSpans = make([]int, maxSpans)
+	c.activeSpansIndex = make([]int, maxSpans)
+	c.openSpans = make([]bool, maxSpans)
+	c.nChildren = make([]int, maxSpans)
+	c.execDuration = make([]float64, maxSpans)
+	c.parentIdx = make([]int, maxSpans)
+	c.timestamps = make(timestampArray, 2*maxSpans)
+}
+
+// reset initializes structures of the sublayer calculator to prepare the computation of sublayer metrics
+// for a trace with max n spans
+func (c *SublayerCalculator) reset(n int) {
+	if c.maxSpans < n {
+		c.initFields(n)
+	}
+	for i := 0; i < n; i++ {
+		c.activeSpansIndex[i] = inactiveSpan
+		c.activeSpans[i] = 0
+		c.openSpans[i] = false
+		c.nChildren[i] = 0
+		c.execDuration[i] = 0
+	}
+}
+
+// a timestamp is the start or end of a span
+// we store the span and parent index for each timestamp instead
+// of storing a pointer to the span to be able to reuse the sublayer calculator arrays
+type timestamp struct {
+	spanStart bool
+	spanIdx   int
+	parentIdx int
+	ts        int64
+}
+
+// timestampArray is an array of timestamp. Used to define Len, Swap and Less to be able to sort the array
+type timestampArray []timestamp
+
+func (t timestampArray) Len() int      { return len(t) }
+func (t timestampArray) Swap(i, j int) { t[i], t[j] = t[j], t[i] }
+
+// for spans with a duration of 0, we need to open them before closing them
+func (t timestampArray) Less(i, j int) bool {
+	return t[i].ts < t[j].ts || (t[i].ts == t[j].ts && t[i].spanStart && !t[j].spanStart)
+}
+
+// buildTimestamps builds puts starts and ends of each span in an array of timestamps, and sorts the array.
+// result if stored in the SublayerCalculator
+func (c *SublayerCalculator) buildTimestamps(trace pb.Trace) {
```

Comment:
None of these methods are ever re-used, and they are always called sequentially, and only once. My opinion is that they spread out code too much and instead we could have just one big central `computeDurations` function. Basically you can absorb `buildTimestamps`, `computeParentIdx` into `computeExecDurations` and rename it `computeDurations`. Each piece could have just an inline comment:
```go
// build all trace timestamps
[...]
// compute parent indexes
[...]
```